libbeat/registry/backend/memlog/diskstore.go

```diff
@@ -0,0 +1,724 @@
+		return
+	}
+
+	_, err = f.Seek(0, os.SEEK_END)
+	if err != nil {
+		f.Close()
+		return
+	}
+
+	if s.logNeedsTruncate {
+		s.logEntries = 0 // reset counter if file was truncated on Open
+	}
+
+	s.logNeedsTruncate = false
+	s.logFile = f
+	s.logBuf = bufio.NewWriterSize(&ensureWriter{s.logFile}, s.bufferSize)
+}
+
+func (s *diskStore) Close() error {
+	if s.logFile != nil {
+		err := s.logFile.Close()
+		s.logFile = nil
+		s.logBuf = nil
+		return err
+	}
+
+	return nil
+}
+
+func (s *diskStore) mustCheckpoint() bool {
+	if s.logInvalid {
+		// log was not 'complete' on last open or most recent transaction failed when writing the log file
+		// -> force checkpoint and clear log file
+		return true
+	}
+
+	// Ensure log file is open. If log file can not be opened, we fall back to
+	// checkpointing.
+	if s.logFile == nil {
+		s.tryOpenLog()
+
+		if s.logFile == nil {
+			return true
+		}
+	}
+
+	return false
+}
+
+func (s *diskStore) numLogs() uint {
+	return s.logEntries
+}
+
+// commitCheckpoint finishes a transaction by writing a new registry data file.
+// The new registry it's name will be `<txid>.json`.
+//
+// The txid will be incremented upon success.
+//
+// 1. Create new checkpoint file
+//   - create/truncate and open tmp file
+//   - Serialize registry table to temporary file
+//   - fsync temporary checkpoint file
+//   - close temporary checkpoint file
+//   - Rename temporary checkpoint file to new `<txid>.json` data file
+//   - fsync on store directory
+// 2. Truncate transaction log
+// 3. Update 'active.json' symlink to most recent data file:
+//   - create temporary symlink file pointing to new `<txid>.json>` data file
+//   - rename temporary symlink file to `active.json`
+//   - fsync on store directory
+// 4. Remove old data files
+func (s *diskStore) commitCheckpoint(tbl *hashtable) error {
+	tmpPath, err := s.checkpointTmpFile(filepath.Join(s.home, "checkpoint"), tbl)
+	if err != nil {
+		return err
+	}
+
+	fileTxID := s.txid + 1
+	fileName := fmt.Sprintf("%v.json", fileTxID)
+	checkpointPath := filepath.Join(s.home, fileName)
+
+	if err := os.Rename(tmpPath, checkpointPath); err != nil {
+		return err
+	}
+	s.syncHome()
+
+	// clear transaction log once finished
+	s.checkpointClearLog()
+
+	// finish current on-disk transaction by increasing the txid
+	s.txid++
+
+	s.dataFiles = append(s.dataFiles, dataFileInfo{
+		path: checkpointPath,
+		txid: fileTxID,
+	})
+
+	// delete old transaction files
+	if err := s.updateActiveSymLink(); err == nil {
+		s.removeOldDataFiles()
+	}
+
+	s.syncHome()
+	return nil
+}
+
+func (s *diskStore) updateActiveSymLink() error {
+	activeLink := filepath.Join(s.home, "active.json")
+
+	active, _ := activeDataFile(s.dataFiles)
+	if active == "" {
+		os.Remove(activeLink) // try, remove active symlink if present.
+		return nil
+	}
+
+	active = filepath.Base(active)
+	tmpLink := filepath.Join(s.home, "active.json.tmp")
+	if err := os.Symlink(active, tmpLink); err != nil {
+		return err
+	}
```

Comment:
Let's keep an eye open for the _Symlink_ behavior, I think it's should be fine for linux or unix system but I vaguely remember some different behavior on windows systems and go version. 

Looking at that specific issue https://github.com/golang/go/issues/22874